### PR TITLE
8267832: SimpleVisitors and Scanners in jdk.compiler should use @implSpec

### PR DESCRIPTION
**Pull request by lahodaj**
As noted in:
https://bugs.openjdk.java.net/browse/JDK-8265981?focusedCommentId=14423316&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14423316

Methods in various utility visitor classes in jdk.compiler should use @implSpec to specify the implementation behavior. This patch tries to add the @implSpec tag to methods which already contain a text specifying the implementation, and adds new javadoc to the handful of methods that are missing it so far.

The CSR is started for review here:
https://bugs.openjdk.java.net/browse/JDK-8267838

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267832](https://bugs.openjdk.java.net/browse/JDK-8267832): SimpleVisitors and Scanners in jdk.compiler should use @implSpec

### Reviewers
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to 5c9f694fade7380fa1a7ae667cf4acd5be60c244
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to 5c9f694fade7380fa1a7ae667cf4acd5be60c244

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4223/head:pull/4223` \
`$ git checkout pull/4223`

Update a local copy of the PR: \
`$ git checkout pull/4223` \
`$ git pull https://git.openjdk.java.net/jdk pull/4223/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4223`

View PR using the GUI difftool: \
`$ git pr show -t 4223`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4223.diff">https://git.openjdk.java.net/jdk/pull/4223.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
  * Inside your method, call super.visitXYZ to visit descendant
  * nodes.
  *
+ * @implSpec
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:38]: 1. Should the error-counting example in the immediately following paragraph belong to this `@implSpec` section?
2. Why is there no similar `@implSpec` section being added to the class-level doc comment of TreeScanner? If such a section is to be added, then should the identifier-counting example in the immediately following paragraph belong to that section?

**Review by pavelrappo — COMMENTED** [review]
I wish we had a better class-level (Doc)TreeScanner doc comment. That would allow to avoid most of the repetition in method-level comments, the vast majority of which are there only to say that the children are scanned in left to right order.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java (hunk 1 above)
Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:52]: Thanks for the comment, Pavel. I've tried to do the changes in an update:
https://github.com/openjdk/jdk/pull/4223/commits/5c9f694fade7380fa1a7ae667cf4acd5be60c244

**Review by jddarcy — APPROVED** [review]

**Review by pavelrappo — APPROVED** [review]

**Pull request closed without merging**